File: 01_DFS/JongYoon/MinStack.java
class MinStack {
    
    class Node{
        int num;
        int min_num;
    }
    
    LinkedList<Node> ll = new LinkedList<>();

    public void push(int x) {
        
        Node node = new Node();
        node.num = x;
        try{
            Node lastNode = ll.getLast();
		    if(lastNode.min_num > x){
			    node.min_num = x;
		    } else {
		        node.min_num = lastNode.min_num;
		    }
        } catch(Exception e){
            node.min_num = x;
        }
	
		ll.add(node);
	}

	public void pop() {
		ll.removeLast();
	}

	public int top() {
		return ll.getLast().num;
	}

	public int getMin() {
		return ll.getLast().min_num;
	}
}

File: 01_DFS/Eunbeom/minStack.cpp
class MinStack {
public:
    stack<int> s, m;
    
    void push(int x) {
        if(m.empty() || x < m.top()){
            m.push(x);
        }else{
            m.push(m.top());
        }
        s.push(x);
    }

    void pop() {
        s.pop();
        m.pop();
    }

    int top() {
        return s.top();
    }

    int getMin() {
        return m.top();
    }
};